### PR TITLE
SDK honors [FunctionName] attribute

### PR DESCRIPTION
**Pull request by MikeStall**
SDK honors [FunctionName] attribute
Resolve https://github.com/Azure/azure-webjobs-sdk/issues/1170
ADd new JobHost.CallAsync(string) endpoint which can honor the method name.

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Protocols/FunctionDescriptor.cs [hunk 1]
```diff
@@ -34,10 +34,14 @@ public class FunctionDescriptor
...
 #if PUBLICPROTOCOL
 #else
-        /// <summary>Gets or sets the name used for logging. This is 'Method'. </summary>
+        /// <summary>Gets or sets the name used for logging. This is 'Method' or the value overwritten by [FunctionName] </summary>
         [JsonIgnore] 
         internal string LogName { get; set; }
 
+        /// <summary>Gets or sets the name used for logging. This is ShortName or the value overwitten by [FunctionName]. </summary>
+        [JsonIgnore] 
+        internal string ShortNameOverride { get; set; }
```
Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.Protocols/FunctionDescriptor.cs]: this is a pretty confusing property: 

1. should it be formatted as `Class.UserOverride`?
2. do we need two separate `ShortName` properties? I think that people will choose incorrectly

File: src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndex.cs [hunk 2]
```diff
@@ -46,6 +46,20 @@ public IFunctionDefinition Lookup(string functionId)
             return _functionsById[functionId];
         }
 
+        public IFunctionDefinition LookupByName(string name)
```
Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndex.cs]: lookup by which name type? Can we make it more explicit from the function name or definition?

File: src/Microsoft.Azure.WebJobs.Protocols/FunctionDescriptor.cs (hunk 1 above)
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Protocols/FunctionDescriptor.cs]: It's confusing ... But I can't find a correct alternative that actually complies with our maze of backwards compat requirements.

Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.Protocols/FunctionDescriptor.cs]: what are the other uses for ShortName?

**Pull request merged**